identity/src/main/java/com/android/identity/credentialtype/CredentialAttributeType.kt

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.identity.credentialtype
+
+/**
+ * Enum of the different types of Credential Attributes
+ */
+sealed class CredentialAttributeType {
+    object STRING : CredentialAttributeType()
+    object NUMBER : CredentialAttributeType()
+    object DATE : CredentialAttributeType()
+    object DATE_TIME : CredentialAttributeType()
+    object PICTURE : CredentialAttributeType()
+    object BOOLEAN : CredentialAttributeType()
+    class StringOptions(val options: List<StringOption>): CredentialAttributeType()
+    class IntegerOptions(val options: List<IntegerOption>): CredentialAttributeType()
+    class ComplexType(val typeName: String, val isArray: Boolean): CredentialAttributeType()
```

Comment:
I'm guessing `ComplexType` is a way to easily support editable `driving_privileges` from the "Add self-signed document" UI? So, I think that's clearly desirable although I also think we actually don't need it (most users of the "Add self-signed document" won't really edit e.g. `driving_privileges`). Some thoughts.

First, I think it should be done in a way so it doesn't "leak" into this system... because the main goal of this system isn't really to support "Add self-signed document" UIs. No, the main goal of the system is to support wallets which need to render consent dialogs and readers which need to render "what to request?" dialogs. It's also a goal to support issuance systems but those systems will pull data from an existing System of Record and they'll know how to craft e.g. `driving_privileges`.

So I think we should simply just have it be `object Complex : CredentialAttributeType()` which is a signal that none of `STRING`, `NUMBER`, `DATE`, `DATE_TIME`, `PICTURE`, `BOOLEAN`, `StringOptions`, `IntegerOptions` apply and the app need special handling if it wants to produce or consume data from that attribute. And then move all the complex processing of the actual type (e.g. `driving_privileges`) into the wallet app itself... I think that could probably be done in a pretty nice way and without the implicit relationships we have here with e.g. `DrivingPrivilege.vehicle_category_code`

Reply:
For example the VehicleRegistration almost does not make sense without complex types. Even the personal data is part of complex types. So I think it should be desirable to have the complete data model in the CredentialTypeRepository and not only the root elements. Also to have a user friendly name for the presentation of the document in a reader app. Maybe the complex type definitions could be a separated list e.g.

```
class MdocNamespace(
    val namespace: String,
    val dataElements: List<MdocDataElement>
    val complexTypeDefinitions: List<MdocChildElement> // or another type with the identifier splitted in typeName and elementName (instead of the dot in the identifier)
)
```
In this way you can use the dataElements in the context of "What to request" and the complexTypeDefinitions for presentation purposes.

Reply:
Ok, just moved the complex types to the holder app. Can this be merged?